crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,406 @@
+use core::{convert::Infallible, ops::AddAssign};
+
+use bitcoin::{OutPoint, Script, Transaction, TxOut};
+
+use crate::{
+    keychain::Balance,
+    sparse_chain::ChainPosition,
+    tx_graph::{Additions, TxGraph, TxNode},
+    BlockAnchor, ChainOracle, FullTxOut, ObservedIn, TxIndex,
+};
+
+/// An outwards-facing view of a transaction that is part of the *best chain*'s history.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TxInChain<'a, T, A> {
+    /// Where the transaction is observed (in a block or in mempool).
+    pub observed_in: ObservedIn<&'a A>,
+    /// The transaction with anchors and last seen timestamp.
+    pub tx: TxNode<'a, T, A>,
+}
```

Comment:
I am a bit confused. The `TxNode` already contains the anchors. Can't we derive `ObservedIn` from the anchors themselves? The same anchor, at least one of them, will also be in `ObservedIn::Block`. If the anchor set is empty, then its a mempool and we already know `last_seen_time` in `TxNode`. 

So why do we need this extra structure?

Also, better to call it a `TxData` or something similar. Its semantically not a node in the graph.

Reply:
`TxNode` has many anchors. This after you've narrowed it down to one. You could say that it's redundant to have the full `TxNode` along with this but I guess it's nice to know what other blocks this tx has shown up in without doing another look up?

Reply:
>> You could say that it's redundant to have the full TxNode along with this but I guess it's nice to know what other blocks this tx has shown up in without doing another look up?

I do think it's a bit redundant to keep track of all the orphan blocks this tx has shown up in, as bitcoin core nodes only reports blockhash of a transaction that is in "active chain". And these block conflicts are resolved by the node itself, so we don't need to handle them ourselves. So eventually we will only work with one anchor per tx. 

So, in theory, we could always maintain the "narrowed down" anchor for a transaction `TxNode` itself, which to me seems would make some of the logic and structure around `TxGraph` a bit simpler?

Reply:
The current implementation is just the most straightforward way of presenting all the data that we have. I don't think there is any benefit for a second structure with most of the fields of `TxNode`, except only supports a single anchor.